src/header/common/cookie.rs

```diff
@@ -59,18 +77,82 @@ impl Header for Cookie {
     }
 }
 
+impl Cookie {
+    /// Create an empty Cookie header.
+    pub fn new() -> Cookie {
+        Cookie::with_capacity(0)
+    }
+
+    /// Create a Cookie header of a certain size.
+    pub fn with_capacity(capacity: usize) -> Cookie {
+        Cookie {
+            cookies: Vec::with_capacity(capacity),
+            index: VecMap::with_capacity(capacity),
+        }
+    }
+
+    /// Shrink the Cookie header internal elements to the currently used size.
+    pub fn shrink_to_fit(&mut self) {
+        self.cookies.shrink_to_fit();
+    }
+
+    /// Append a new cookie to the Cookie header.
+    pub fn push<T: Into<String>>(&mut self, name_tref: T, value_tref: T) {
+        let name = name_tref.into();
+        let value = value_tref.into();
+        self.cookies.push((name.clone(), value.clone()));
+        if self.index.get(&name) == None {
+            self.index.insert(name, value);
+        }
+    }
+
+
+    /// Get value of cookie from name. If duplicate names were pushed to the
+    /// Cookie header, this function will only return the first one.
+    pub fn get<T: Borrow<String>>(&self, name: T) -> Option<String> {
+        match self.index.get(name.borrow()) {
+            Some(value_ref) => Some((*value_ref).clone()),
+            None => None,
+        }
+    }
+
+
+    /// Clear the current Cookie, and add one with specified name and value.
+    pub fn set<T: Into<String>>(&mut self, name_tref: T, value_tref: T) {
+        self.cookies.clear();
+        self.index.clear();
+        self.push(name_tref, value_tref);
+    }
+
+    /// Check if there are any exiting cookie.
+    pub fn is_empty(&self) -> bool {
+        self.cookies.is_empty()
+    }
+}
+
 impl fmt::Display for Cookie {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let cookies = &self.0;
-        for (i, cookie) in cookies.iter().enumerate() {
-            if i != 0 {
-                try!(f.write_str("; "));
+        let mut first = true;
+        for pair in self.cookies.clone() {
+            let cookie = format!("{}={}", pair.0, pair.1);
```

Comment:
This would allocate a brand new `String`, before then writing it to the formatter. It can be done without the temporary allocated, such as `write!(f, "{}={}", key, val)`.